### PR TITLE
Use parent pom 4.32, require Jenkins 2.289.1 or newer

### PR DESCRIPTION
**Pull request by MarkEWaite**
## Use parent pom 4.32, require Jenkins 2.289.1 or newer

Use the most recent parent pom.  It provides spotbugs analysis and requires Java 8 syntax.

https://stats.jenkins.io/pluginversions/clover.html shows that 90% of the 1321 installations of clover plugin 4.12.1 are on Jenkins 2.289.1 or newer.  Users that are updating the plugin version are also updating their Jenkins version.

Added some code changes to resolve spotbugs warnings and added one suppression to silence a spotbugs warning.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -61,29 +60,29 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins.workflow</groupId>
             <artifactId>workflow-aggregator</artifactId>
-            <version>2.0</version>
+            <version>2.6</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
```
Comment (marek-parfianowicz) [inline pom.xml:66]: What are `script-security`, `mailer` and `symbol-annotation` used for?

File: src/main/java/hudson/plugins/clover/CloverBuildWrapper.java [hunk 2]
```diff
@@ -27,8 +27,10 @@
 import org.apache.commons.lang.StringUtils;
 import org.kohsuke.stapler.DataBoundConstructor;
 import org.kohsuke.stapler.StaplerRequest;
+import edu.umd.cs.findbugs.annotations.NonNull;
```
Comment (marek-parfianowicz) [inline src/main/java/hudson/plugins/clover/CloverBuildWrapper.java:30]: Why did you change `javax.annotation` to `edu.umd.cs.findbugs.annotations`? The former one was JDK standard, the latter comes from a third party library.

File: pom.xml [hunk 3]
```diff
@@ -30,9 +30,8 @@
     </scm>
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <!-- from org.jenkins-ci.plugins:plugin 3.x -->
-        <!--<java.level>8</java.level>-->
-        <!--<java.level.test>8</java.level.test>-->
+        <java.level>8</java.level>
+        <jenkins.version>2.289.1</jenkins.version>
```
Comment (marek-parfianowicz) [inline pom.xml:34]: > shows that 90% of the 1321 installations of clover plugin 4.12.1 are on Jenkins 2.289.1 or newer. 

I agree, we can bump it.

File: pom.xml [hunk 4]
```diff
@@ -30,9 +30,8 @@
     </scm>
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <!-- from org.jenkins-ci.plugins:plugin 3.x -->
-        <!--<java.level>8</java.level>-->
-        <!--<java.level.test>8</java.level.test>-->
+        <java.level>8</java.level>
```
Comment (marek-parfianowicz) [inline pom.xml:33]: Having a minium requirement of Java 8 is fine. Java 7 has marginal use.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java [hunk 5]
```diff
@@ -157,7 +157,8 @@ private void performImpl(Run<?, ?> run, FilePath workspace, TaskListener listene
             }
 
             // if the run has failed, then there's not much point in reporting an error
-            final boolean buildFailure = run.getResult() != null && run.getResult().isWorseOrEqualTo(Result.FAILURE);
+            Result result = run.getResult();
```
Comment (marek-parfianowicz) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java:160]: Add `final` for consistency, maybe?

File: src/main/java/hudson/plugins/clover/CloverPublisher.java [hunk 6]
```diff
@@ -272,6 +273,10 @@ private boolean copyHtmlReport(FilePath coverageReport, FilePath buildTarget, Ta
             return false;
         }
         final FilePath htmlDirPath = htmlIndexHtmlPath.getParent();
+        if (htmlDirPath == null) {
```
Comment (marek-parfianowicz) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java:276]: Is this possible to have a directory without a parent? Only for `/`, I guess? But is it even possible to have such Jenkins configuration?

File: pom.xml (hunk 1 above)
Comment (MarkEWaite) [reply pom.xml:66]: Just to satisfy compilation and upper bounds dependencies.  They will be removed in a future change that is ready in my repository.  

When the plugin bill of materials is used, those unexpected dependencies are no longer required.

File: src/main/java/hudson/plugins/clover/CloverBuildWrapper.java (hunk 2 above)
Comment (MarkEWaite) [reply src/main/java/hudson/plugins/clover/CloverBuildWrapper.java:30]: A comment in [JENKINS-55973](https://issues.jenkins.io/browse/JENKINS-55973) describes why JSR-305 has been replaced with spotbugs annotations in Jenkins core and in Jenkins plugins.  

The fundamental message (for me) was that JSR-305 was never completed.  It never became a part of the officially supported JDK.  Thankfully, Google and others have adopted spotbugs as the actively developed replacement for findbugs.  It is maintained and often releases new versions with additional capabilities.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java (hunk 6 above)
Comment (MarkEWaite) [reply src/main/java/hudson/plugins/clover/CloverPublisher.java:276]: I doubt that it is possible to have a directory without a parent.  I didn't want to wrestle with the spotbugs warning that appears if `htmlDirPath` is not checked for null.  It seemed easier to check for null than to insert a spotbugs suppression.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java (hunk 5 above)
Comment (MarkEWaite) [reply src/main/java/hudson/plugins/clover/CloverPublisher.java]: Good suggestion.  Done in 7905e1e

File: pom.xml (hunk 1 above)
Comment (MarkEWaite) [reply pom.xml:66]: If you'd prefer to not have the surprising dependency temporarily, I can update this pull request with the plugin bill of materials so that the dependencies are easier to understand.  The plugin bill of materials change is a larger change because it also adds a declarative Pipeline test and creates a `clover` keyword to be used in Jenkins Pipeline.

Comment (MarkEWaite) [reply pom.xml:66]: See https://github.com/jenkinsci/clover-plugin/compare/master...MarkEWaite:use-plugin-bom for the `use-plugin-bom` commit that is added onto this branch.  It simplifies the test dependencies, adds a declarative Pipeline test, and adds the `clover` symbol that enables [Pipeline syntax snippet generator](https://www.jenkins.io/doc/book/pipeline/getting-started/#snippet-generator).  

Unfortunately, the snippet generator seems to ignore some of the entered values from the clover plugin fields.  I don't understand why those field values are ignored, but using the snippet generator is still easier (at least for me) than entering the syntax interactively.

File: src/main/java/hudson/plugins/clover/CloverBuildWrapper.java (hunk 2 above)
Comment (marek-parfianowicz) [reply src/main/java/hudson/plugins/clover/CloverBuildWrapper.java:30]: Thank you for clarification. Let's switch to findbugs-annotations then.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java (hunk 6 above)
Comment (marek-parfianowicz) [reply src/main/java/hudson/plugins/clover/CloverPublisher.java:276]: OK

File: pom.xml (hunk 1 above)
Comment (marek-parfianowicz) [reply pom.xml:66]: > I can update this pull request with the plugin bill of materials so that the dependencies are easier to understand. The plugin bill of materials change is a larger change.

Let's keep changes small, we can have these dependencies as an intermediate step, especially that they're in the test scope.

**Review by marek-parfianowicz — APPROVED** [review]
